### PR TITLE
Adding 'clone to satisfy the borrow checker' anti-pattern

### PR DESCRIPTION
**Pull request by liamzdenek**
I feel like this write-up could be improved, but I'm not exactly sure how -- recommendations and modifications are welcome.

### CODE REVIEW

File: anti_patterns/borrow_clone.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+The borrow checker prevents rust users from developing otherwise unsafe code by
+ensuring that either: only one mutable reference exists, or potentially many but
+all immutable references exist. If the code written does not hold true to these 
+conditions, this anti-pattern arises when the developer resolves the compiler
+error by cloning the variable.
+
+
+## Example
+
+```rust fn main() {
```
Comment (lfairy) [inline anti_patterns/borrow_clone.md]: The "rust" and "fn main()" should be on separate lines, like this:

    ```rust
    fn main() {

File: anti_patterns/borrow_clone.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Clone to satisfy the borrow checker
+
+## Description
+
+The borrow checker prevents rust users from developing otherwise unsafe code by
```
Comment (lfairy) [inline anti_patterns/borrow_clone.md]: "Rust" should be capitalized

File: anti_patterns/borrow_clone.md (hunk 1 above)
Comment (lfairy) [inline anti_patterns/borrow_clone.md]: Also, I don't think you need to wrap the code in `fn main()` -- the other examples in this repo (along with the official docs) keep the code bare.

File: anti_patterns/borrow_clone.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+// perform some action on the borrow to prevent rust from optimizing this
+//out of existence
+*y += 1;
+
+// without the x.clone() two lines prior, this line would fail on compile as
+// x has been borrowed
+// thanks to x.clone(), x was never borrowed, and this line will run.
+println!("{}", x);
+```
```
Comment (cbreeden) [inline anti_patterns/borrow_clone.md]: I think we can come up with a better example; I'm not sure if this is really a problem when your type is copy.  Let me play around with some examples and get back to you.

**Comment by cbreeden** [comment]
I like your motivation and description for the most part.  I think we can make improve this section quite a bit if we put a little more polish into the example, and add examples that shows how you can avoid the problems.  I'll play around with a few ideas and get back to you.

**Comment by cbreeden** [comment]
So I tried to think of some good examples, what do you think about this one

```
    let mut vec = vec![("Bob".to_string(), 17), ("Stacy".to_string(), 21)];
    let new_pair = ("Ferris".to_string(), 13);
    
    // check for duplicates
    let (ref key, _) = new_pair;
    if vec.iter().map(|&(ref k, _)| k).find(|&k| k == key).is_none() {
        // Can't move since new_pair is being referenced still!
        vec.push(new_pair);
    }
    
    println!("{:?}", vec);
```

I could imagine someone trying to add a `.clone()` to the `new_pair` in `vec.push(new_pair);`.  A lot of times these problems can be solved by properly managing their lifetimes -- in this case just adding `k == &new_pair.0`.  In other cases, closing a mutable borrow scope in brackets.  I don't know if this is the best example, just throwing it out there.

**Comment by llogiq** [comment]
I have a few ideas for examples:

* change an enum value, leaving parts intact (see #31)
* select & remove items from a collection (use Drain instead once available)
* un-by-ref an iterator (`.cloned()`) which is actually OK in many cases (unless you can consume the collection while iterating, which should be preferred)
* `entry(_.clone())` – there's an RFC to get rid of that.

Hope that helps.

**Comment by pickfire** [comment]
Anything left for this?

**Comment by simonsan** [comment]
@liamzdenek any updates on this?

**Comment by simonsan** [comment]
Closing due to inactivity. Further additions to the PR can be made in #110.

**Pull request closed without merging**